File: baekjoon/9466.js
// 텀 프로젝트 : 그래프, 깊이 우선 탐색, 해시를 사용한 집합과 맵
const input = require("fs")
  .readFileSync("/dev/stdin")
  .toString()
  .trim()
  .split("\n")
  .map((v) => v.split(" ").map((v) => Number(v)));
const [caseCnt] = input.shift();
let inputIndex = 0;
let i = 0;
const answer = [];

while (i < caseCnt) {
  const [studentCnt] = input[inputIndex++];
  const favoriteArr = input[inputIndex++];
  const result = new Array(studentCnt + 1).fill(0);
  for (let j = 1; j <= studentCnt; j++) {
    const trace = [j];
    const visited = {};
    let traceIndex = 0;
    visited[j] = traceIndex;

    while (result[j] === 0) {
      const next = favoriteArr[trace[trace.length - 1] - 1];
      if (result[next] != 0) {
        for (let k = 0; k < trace.length; k++) {
          result[trace[k]] = 1;
        }
        break;
      } else if (visited[next]) {
        let startIndex = visited[next];
        for (let k = 0; k < trace.length; k++) {
          if (k >= startIndex && k <= traceIndex) {
            result[trace[k]] = 2;
          } else {
            result[trace[k]] = 1;
          }
        }
        break;
      } else {
        trace.push(next);
      }
      visited[next] = ++traceIndex;
    }
  }
  let count = 0;
  for (let i = 1; i < result.length; i++) {
    if (result[i] == 1) {
      count++;
    }
  }
  answer.push(count);
  i++;
}

console.log(answer.join("\n"));

// 메모리 초과
// - 잘못된 로직에 의해서도 메모리 초과가 발생할 수 있다
// - e.g. (무한 반복에 가까운) 많은 반복이 발생하여 그 정보를 모두 저장하는 경우
// - 시간 초과가 발생하기 전에 메모리 초과가 발생할 수 있다(자주 발생할 것 같지는 않다)

// 시간 초과
// - result 배열에 저장하지 않은 노드에서 result 배열에 저장한 노드로 접근할 수도 있다
// - 이 경우 똑같은 로직을 반복하기 떄문에 시간 초과가 발생할 수 있다

File: baekjoon/9466.cpp
// 텀 프로젝트 : 그래프, 깊이 우선 탐색, 해시를 사용한 집합과 맵
#include <vector>
#include <iostream>
#include <map>

using namespace std;

int main() {

	typedef std::vector<std::pair<int, std::vector<int> > > InputVector;

	int caseCnt;
	int i = 0;
	InputVector 		input;
	std::vector<int>	answer;

	cin >> caseCnt;

	while (i < caseCnt) {
		int j = 0;
		int studentCnt;
		std::vector<int> favoriteArr;
		cin >> studentCnt;
		while (j < studentCnt) {
			int tmp;
			cin >> tmp;
			favoriteArr.push_back(tmp);
			j++;
		}
		input.push_back(make_pair(studentCnt, favoriteArr));
		i++;
	}
	i = 0;

	while (i < caseCnt) {
		int					studentCnt = input[i].first;
		std::vector<int>	favoriteArr = input[i].second;
		std::vector<int>	result(studentCnt + 1, 0);
		for (int j = 1; j <= studentCnt; j++) {
			std::vector<int>	trace;
			std::map<int, int>	visited;
			int					traceIndex = 0;

			trace.push_back(j);
			visited[j] = traceIndex;

			while (result[j] == 0) {
				int next = favoriteArr[trace[trace.size() - 1] - 1];
				if (result[next] != 0) {
					for (int k = 0; k < trace.size(); k++) {
						result[trace[k]] = 1;
					}
					break;
				} else if (visited.find(next) != visited.end()) {
					int startIndex = visited.find(next)->second;
					for (int k = 0; k < trace.size(); k++) {
						if (k >= startIndex && k <= traceIndex) {
							result[trace[k]] = 2;
						} else {
							result[trace[k]] = 1;
						}
					}
					break;
				} else {
					trace.push_back(next);
				}
				visited[next] = ++traceIndex;
			}
		}
		int count = 0;
		for (int i = 1; i < result.size(); i++) {
			if (result[i] == 1) {
				count++;
			}
		}
		answer.push_back(count);
		i++;
	}
	for (int i = 0; i < answer.size(); i++) {
		cout << answer[i] << endl;
	}
}

// 시간 초과
// 12
// 2 3 4 5 6 7 8 9 10 10 2 2

// result[i] == 0 인지를 확인하고 trace 배열을 채워나감
// 한 번 방문한 노드는 다시 방문하지 않아야 함 => 방문하면 시간 초과가 발생
// 1. trace 배열을 처음 채울 때 result 배열을 채우기 때문에, 이 때 채워진 노드는 다시 방문하지 않음
// 2. 하지만 이전 단계에서 채워지지 않은 노드가 이미 방문한 노드를 가리키는 경우도 있음